On working with a ViewModel Composition based system (#262)
mauroservienti/milestone.github.io: _posts/on-working-with-viewmodel-composition-based-system.md

        AH --> VM[ViewModel]
        BH --> VM[ViewModel]
    end
    subgraph Browser
        VM-->F[[Frontend]]
    end
</div>

If we're building a distributed system, we're probably also following, more or less intentionally, the [4+1 architecture view model](https://en.wikipedia.org/wiki/4%2B1_architectural_view_model).

The presented diagram is a mix of the physical and process views. It's strictly neither of the two. However, it's useful for our discussion because, in some way, it's the final state: Whatever we're doing, the system, when deployed in production, needs to look like that.

The logical view, however, is quite different:

<div class="mermaid">
graph LR
    subgraph Logical service A
        A((Data))
        A --> AH[Composition Handler A]
    end
    subgraph Logical service B
        B((Data))
        B --> BH[Composition Handler B]
    end
    subgraph Branding service
        AH --> VM[ViewModel]
        BH --> VM
        VM-->F[Presentation]
    end
</div>

> I'm intentionally skipping UI Composition/micro-frontends. It can only be more complex. Please, leave a comment below if you think UI Composition aspects need any further in-depth analysis.

> In between the logical and physical/process view of the system sits the development view. The development view enables the logical view to get to life. At the same time, it mediates and adapts from the logical view to the runtime needs respecting the non-functional requirements imposed by the tool-chain in use.

Reviewer: I switched "needs" to "requirements" because I read "needs" as a verb and couldn't follow the syntax. I don't really like using "requirements" twice in the same sentence but I think it's okay here.

```suggestion
In between the logical and physical/process view of the system sits the development view. The development view enables the logical view to come to life. At the same time, it mediates and adapts from the logical view to the runtime requirements respecting the non-functional requirements imposed by the tool-chain in use.
```

To recap, if we're going down the route of a single repository for each logical service, there will be pros and cons:

- Pros: the overall development process will be smoother
- Pros: multiple components integration testing will be straightforward
- Cons: more complex CI/CD pipelines
- Cons: each release will release all components, even though some might be released with no changes

What else can we do?

## A repository per component 

We could have a different repository for each component: A repository for the restful APIs, one for each composition handler, another for the composed ViewModel, and a last one for the frontend.

CI/CD pipelines would be much easier. The restful API can be deployed independently from other components, and the same goes for all the components. Composition handlers could be deployed as Nuget packages, consumed by the composition gateway host at deployment time. Composition handlers and the fronted would also depend on the ViewModel package.

> By doing so, we'd get fine-grained control over packages deployment and versioning. It would be much easier to set up and customize CI/CD pipelines for each component. And we won't waste any version number because each component would only be released when strictly necessary.

Reviewer: ```suggestion
By doing so, we'd get fine-grained control over package deployment and versioning. It would be much easier to set up and customize CI/CD pipelines for each component. And we won't waste any version numbers because each component would only be released when strictly necessary.
```

> That begin said, it's not gold all that glitters. Setting up integration testing is way more complicated when using different repositories for each component. It's not impossible. However, we'll have to spend time designing custom components and tools specifically aimed at supporting integration tests.

Reviewer: ```suggestion
That being said, all that glitters is not gold. Setting up integration testing is way more complicated when using different repositories for each component. It's not impossible. However, we'll have to spend time designing custom components and tools specifically aimed at supporting integration tests.
```

Last but not least. Even if it's true that using a repository for each component is a cleaner approach, it'll introduce friction into the development process, especially at the early stages. Let's imagine that we're adding a new restful API. The first step is to change the restful API repository; then, we need to publish the changes so that downstream components can consume them. The development effort then moves onto composition handlers. Update the dependency on the restful API package to have integration tests fail or be able to create the new required tests. Adjust the composition handlers to support the new restful API and publish the package. Apply, if needed, a similar process to the frontend component.

> > For more details about integration testing when using HTTP endpoints, take a look at ["You wanna test HTTP, right?"](https://milestone.topics.it/2021/04/28/you-wanna-test-http.html) and the latest [HTTP client and HTTP client factory in integration testing](link).

Reviewer: Reminder not to forget the link

I wouldn't define the described process as smooth.

As for the previous approach, let's take a look at the pros and cons:

> - Cons: the overall development process will come with more frictions

Reviewer: ```suggestion
- Cons: the overall development process will have more friction
```

> - Cons: multiple components integration testing will be more challenging

Reviewer: ```suggestion
- Cons: integration testing for multiple components will be more challenging
```

- Pros: each release will release only the changed components

Sounds like exactly the opposite of a repository for each logical service. 

From the structure perspective, a few other approaches might sit in the middle of what we described so far. I guess that they would also lead to a similar distribution of pros and cons.

## Shall we blindly use ViewModel Composition?

The quick answer is obviously no. And not only for the technical challenges presented in this article.

There are many factors that we should consider before deciding to use ViewModel Composition. For example, the company size, team competencies, and the system's overall complexity are all good metrics to look at before making any final decision. ViewModel Composition will only make things more complex if the company or project is small and the team is not so experienced. It'd be better to design frontends using monolithic architecture. One frontend repository contains all the code and artifacts and knows how to communicate with the distributed systems backend.

It's probably fair to say that a composite UI is the last of the problems; not necessarily the least important, but the last to address.

> We tend to think in terms of consequences. The reasoning usually goes like this: if we design that component in a suboptimal way when we'll fix it, it will hurt all these other components.

Reviewer: ```suggestion
We tend to think in terms of consequences. The reasoning usually goes like this: if we design that component in a suboptimal way, then when we fix it, it will hurt all these other components.
```
